rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,180 @@
+  a host scalar?
+    - Currently, we don't have any engine specified to input memory descriptors.
+- How to pass host-side scalars to the execute function?
+  - Currently, we only take dnnl::memory objects.
+
+## How to specify an input as a host-side scalar value to primitive desc creation?
+
+### Option 1: do nothing and take host-side scalars as primitive descriptor parameters
+
+Here user would pass the host side scalar as a parameter to primitive
+creation, as is done today. The main caveats here are:
+- It requires user to recreate primitive descriptors and primitives
+  every time the host-side scalar changes value.
+- If the implementation jits a kernel that depends on that parameter,
+  we can get large overheads due to re-jitting.
+
+Pros:
+- oneDNN API can already handle this for parameters that do not
+frequently change.
+
+Cons:
+- Forces user to recreate primitive descriptor and primitive objects
+  every-time that host-side scalar value changes.
+- Potential extra jitting overhead if internal implementation considers that parameter constant.
+
+This option is not recommended if the host-side scalar value is
+expected to frequently change.
+
+### Option 2 (recommended): expose a new memory descriptor kind
+
+The idea here is to:
+- expose host-side scalars through a new memory descriptor create function
+- have user pass a memory with cpu engine to execute function.
+
+The new kind of memory descriptor would be created through a dedicated function
+It would be very lightweights and be limited:
+- only scalar value, so no `ndims`, no `dims`, ...
+- only scalar datatype can be configurable by user.
+
+```C
+dnnl_status_t DNNL_API dnnl_memory_desc_create_host_scalar(
+        dnnl_memory_desc_t *memory_desc, dnnl_data_type_t data_type);
+```
+
+Internal considerations: we would use a new `format_kind` with no
+associated `format_desc`. This would avoid aliasing host-side scalars
+with blocked descriptors, and hence host-side scalars will need to be
+explicitly supported by implementations.
+
+For users passing primitive objects around and querying them, it makes
+sense to expose this format kind, so users can query if a given input
+is a host-side scalar, and pass it accordingly to primitive execution.
+
+Pros:
+- Avoids user from recreating primitives when scalars are changed.
+- Simple enabling internally.
+- Allows passing scalar as a kernel argument, that should not introduce
+a performance penalty.
+
+Cons:
+- Ties the memory descriptor to a specific device (host) which we
+  avoided so far.
+
+> Note: Options 2 and 3 formally require creating memory objects with a `CPU` engine.
+However, since this is a very specific use case that does not otherwise
+require full CPU support in oneDNN, it is suggested to introduce a new `host` or "null" engine.
+This would allow users to avoid building oneDNN with complete CPU support just for this scenario.
+
+### Option 3: explicitly take engine kind in memory desc
```

Comment:
Formally, our specification [states](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onednn/source/data_model/memory/objects#descriptors) that memory descriptors are engine agnostic so unless we are fine with changing the current data model this option doesn't look viable.

Reply:
yes correct, this option is to open discussion about reevaluating memory descriptors being engine agnostic.

Reply:
resolving as this seems answered